src/replication/replication_node_manager.cpp

```diff
@@ -61,44 +75,75 @@ namespace cubreplication
 
     void finalize ()
     {
+      wait_ha_tasks ();
+
       g_hostname.clear ();
 
       delete g_slave_node;
       g_slave_node = NULL;
       delete g_master_node;
       g_master_node = NULL;
 
-      // stream and stream file are interdependent, therefore first stop the stream
+      // we need to first stop the stream before destroying stream_file
       g_stream->stop ();
       delete g_stream_file;
       g_stream_file = NULL;
       delete g_stream;
       g_stream = NULL;
     }
 
-    void commute_to_master_state ()
+    void commute_to_master_state (const std::function<void (void)> &ha_transitions)
     {
-      delete g_slave_node;
-      g_slave_node = NULL;
-
-      if (g_master_node == NULL)
-	{
-	  g_master_node = new master_node (g_hostname.c_str (), g_stream, g_stream_file);
-	}
+      inc_tasks ();
+
+      cubthread::entry_task *promote_task = new cubthread::entry_callable_task ([&ha_transitions] (
+		  cubthread::entry &context)
+      {
+	if (g_slave_node != NULL)
+	  {
+	    g_slave_node->finish_fetch_from_stream ();
```

Comment:
does it need to wait for fetch from stream only? I think the condition here is to apply all replication.

Ofcourse, part of applying all replication is making sure everything is fetched.

Reply:
Destructor of dispatcher does make sure everything is applied.

Reply:
I understand. Even so, I have two objections about the approach:

  1. just by reading this code, it is not obvious that commute waits for all apply to be done. it is more clear if it calls `g_slave_node->finish_replication_apply ()` (please choose the best name you consider).
  2. I don't think dispatcher destructor is supposed to include applying all replication. it should only make sure resources are properly destroyed; the decision to apply all replication or to abort it must be directed from higher level.

note: if force is true, replication apply should be interrupted; maybe by calling `g_slave_node->abort_replication_apply ()` (or an appropriate name for a function that interrupts replication apply and waits for it to be safely aborted).

Reply:
Waiting applies to finish was easier and is default behavior. For forcefully aborting applying transactions I've opened that http://jira.cubrid.org/browse/CBRD-23178